fabtests/component/sock_test.c

```diff
@@ -48,338 +48,380 @@
 
-static int start_server()
+static struct pollfd *poll_set;
+static struct epoll_event *ep_events;
+static int epfd = -1;
+
+
+static int start_server(void)
 {
 	struct addrinfo *ai = NULL;
-	int ret;
 	int optval = 1;
+	int ret;
 
 	ret = getaddrinfo(opts.src_addr, opts.src_port, NULL, &ai);
 	if (ret) {
-		printf("getaddrinfo failed\n");
-		goto out;
+		FT_PRINTERR("getaddrinfo", ret);
+		return -ret;
 	}
 
 	listen_sock = socket(ai->ai_family, SOCK_STREAM, 0);
 	if (listen_sock < 0) {
-		printf("socket error: %i: %s\n", errno, strerror(errno));
-		ret = -1;
-		goto out;
+		FT_PRINTERR("socket", -errno);
+		goto free;
 	}
 
-	ret = setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, 
-		sizeof(optval));
+	ret = setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR,
+			 (char *) &optval, sizeof(optval));
 	if (ret) {
-		printf("error setting socket options\n");
-		goto out;
+		FT_PRINTERR("setsockopt", -errno);
+		goto close;
 	}
 
 	ret = bind(listen_sock, ai->ai_addr, ai->ai_addrlen);
 	if (ret) {
-		printf("bind failed\n");
-		goto out;
+		FT_PRINTERR("bind", -errno);
+		goto close;
 	}
 
-	ret = listen(listen_sock, connections);
+	ret = listen(listen_sock, 0);
 	if (ret) {
-		printf ("listen failed\n");
-		goto out;
+		FT_PRINTERR("listen", -errno);
+		goto close;
 	}
 
 	freeaddrinfo(ai);
-	return ret;
+	return 0;
 
-out:
+close:
 	close(listen_sock);
+free:
 	freeaddrinfo(ai);
-	return ret;
+	return -errno;
 }
 
-static int server_connect()
+static int server_connect(void)
 {
-	int new_sock, i;
+	int i, ret;
+
+	ret = start_server();
+	if (ret)
+		return ret;
 
 	for (i = 0; i < connections; i++) {
-		new_sock = accept(listen_sock, NULL, NULL);
-		if (new_sock < 0) {
-			printf("error during server init\n");
-			return -1;
+		fds[i] = accept(listen_sock, NULL, NULL);
+		if (fds[i] < 0) {
+			FT_PRINTERR("accept", -errno);
+			ret = -errno;
+			goto close;
 		}
-
-		clients[i] = new_sock;
-		
 	}
+	close(listen_sock);
 	return 0;
+
+close:
+	while (i--) {
+		close(fds[i]);
+		fds[i] = -1;
+	}
+	close(listen_sock);
+	return ret;
 }
 
-static int client_connect()
+static int run_client(void)
 {
 	struct addrinfo *res;
-	int i, ret, new_sock;
+	int i, ret;
+	char c;
 
 	ret = getaddrinfo(opts.dst_addr, opts.dst_port, NULL, &res);
 	if (ret) {
-		printf("getaddrinfo failed\n");
-		goto out;
+		FT_PRINTERR("getaddrinfo", ret);
+		return -ret;
 	}
 
 	for (i = 0; i < connections; i++) {
+		fds[i] = socket(res->ai_family, SOCK_STREAM, 0);
+		if (fds[i] < 0) {
+			FT_PRINTERR("socket", -errno);
+			goto close;
+		}
 
-		new_sock = socket(res->ai_family, SOCK_STREAM, 0);
-		if (new_sock < 0) {
-			ret = -1;
-			goto out;
+		ret = connect(fds[i], res->ai_addr, res->ai_addrlen);
+		if (ret) {
+			FT_PRINTERR("connect", -errno);
+			ret = -errno;
+			goto close;
 		}
+	}
 
-		ret = connect(new_sock, res->ai_addr, res->ai_addrlen);
-		if (ret)
-			goto out;
+	/* wait for server to finish */
+	recv(fds[0], &c, 1, 0);
```

Comment:
Don't we need a minimum error handling here? For example if recv() is interrupted with `EINTR`?

Reply:
Probably.  Let me run this in a debugger and see what's needed.

Reply:
Running in a debugger, I didn't hit any issues.  So, while I could check for an error here, I'm not sure it'll help the test.  I'll at least add a print to report why the client exited early.